contracts/WavePortal.sol

```diff
@@ -4,17 +4,69 @@ pragma solidity ^0.8.17;
 import "hardhat/console.sol";
 
 contract WavePortal {
-    // State Variables
-    uint256 private totalWaves;
+    // Events
+    event NewWave(address indexed from, uint256 timestamp, string message);
+
+    // Structs
+    struct Wave {
+        address wave;
+        string message;
+        uint256 timestamp;
+    }
 
     // Mappings
     mapping(address => uint256) public wavers;
+    mapping(address => uint256) public lastWavedAt;
+
+    // State Variables
+    uint256 private totalWaves;
+    Wave[] waves;
+    uint256 private seed;
+
+    // Constructor
+    constructor() payable {
+        console.log("We have been constructed!");
+
+        // Set initial "random" number for winner
+        seed = (block.timestamp + block.difficulty) % 100;
+    }
 
     // Functions
-    function wave() public {
+    function wave(string memory _message) public {
         wavers[msg.sender] += 1;
         totalWaves += 1;
-        console.log("%s has waved!", msg.sender);
+        console.log("%s has waved and said '%s'", msg.sender, _message);
+
+        // Require waver to not have waved in the last 15 minutes
+        require(
+            lastWavedAt[msg.sender] + 15 minutes < block.timestamp,
+            "Must wait 15 minutes before you can send another message."
+        );
+
+        // Update current senders timestamp
+        lastWavedAt[msg.sender] = block.timestamp;
+
+        // Store wave in waves array
+        waves.push(Wave(msg.sender, _message, block.timestamp));
+
+        // Create new chance of winning for next user
+        seed = (block.timestamp + block.difficulty + seed) % 100;
```

Comment:
You should be aware that there is nothing called Random in blockchain. Everything is Pseudorandom.

Reply:
Okay, thank you for the clarification. I will make sure that future wording in my code clarifies that distinction.